framework/src/timesteppers/Stepper.C

```diff
@@ -0,0 +1,748 @@
+  _solve_time_secs = si._solve_time_secs;
+
+  _soln_nonlin.reset(si._soln_nonlin->clone().release());
+  _soln_aux.reset(si._soln_aux->clone().release());
+  _soln_predicted.reset(si._soln_predicted->clone().release());
+
+  _snapshot = false;
+  _rewind = false;
+  _rewind_time = -1;
+}
+
+StepperInfo&
+StepperInfo::operator=(const StepperInfo& si)
+{
+  _step_count = si._step_count;
+  _time = si._time;
+  _dt = si._dt;
+  _nonlin_iters = si._nonlin_iters;
+  _lin_iters = si._lin_iters;
+  _converged = si._converged;
+  _solve_time_secs = si._solve_time_secs;
+
+  _soln_nonlin.reset(si._soln_nonlin->clone().release());
+  _soln_aux.reset(si._soln_aux->clone().release());
+  _soln_predicted.reset(si._soln_predicted->clone().release());
+
+  _snapshot = false;
+  _rewind = false;
+  _rewind_time = -1;
+  return *this;
+}
+
+void
+StepperInfo::pushHistory(Real dt, bool converged, Real solve_time)
+{
+    _dt.push_front(dt);
+    _dt.pop_back();
+    _converged.push_front(converged);
+    _converged.pop_back();
+    _solve_time_secs.push_front(solve_time);
+    _solve_time_secs.pop_back();
+}
+
+void
+StepperInfo::update(
+    int step_count,
+    Real time,
+    Real dt,
+    unsigned int nonlin_iters,
+    unsigned int lin_iters,
+    bool converged,
+    Real solve_time_secs,
+    std::vector<Real> soln_nonlin,
+    std::vector<Real> soln_aux,
+    std::vector<Real> soln_predicted
+    )
+{
+  _step_count = step_count;
+  _time = time;
+  _dt.push_front(dt);
+  _dt.pop_back();
+  _nonlin_iters = nonlin_iters;
+  _lin_iters = lin_iters;
+  _converged.push_front(converged);
+  _converged.pop_back();
+  _solve_time_secs.push_front(solve_time_secs);
+  _solve_time_secs.pop_back();
+
+  _snapshot = false;
+  _rewind = false;
+  _rewind_time = -1;
+
+  if (_soln_nonlin->size() != soln_nonlin.size())
+  {
+    _soln_nonlin.reset(NumericVector<Number>::build(_dummy_comm).release());
+    _soln_nonlin->init(soln_nonlin.size(), soln_nonlin.size(), false, SERIAL);
+  }
+  if (_soln_aux->size() != soln_aux.size())
+  {
+    _soln_aux.reset(NumericVector<Number>::build(_dummy_comm).release());
+    _soln_aux->init(soln_aux.size(), soln_aux.size(), false, SERIAL);
+  }
+  if (_soln_predicted->size() != soln_predicted.size())
+  {
+    _soln_predicted.reset(NumericVector<Number>::build(_dummy_comm).release());
+    _soln_predicted->init(soln_predicted.size(), soln_predicted.size(), false, SERIAL);
+  }
+
+  *_soln_nonlin = soln_nonlin;
+  *_soln_aux = soln_aux;
+  *_soln_predicted = soln_predicted;
+}
+
+int
+StepperInfo::stepCount()
+{
+  return _step_count;
+}
+
+Real
+StepperInfo::time()
+{
+  return _time;
+}
+
+template <typename T>
+T nth_elem(std::list<T>& lst, int n)
+{
+  int i = 0;
+  for (auto it = lst.begin(); it != lst.end(); ++it)
+  {
+    if (i == n)
+      return *it;
+    i++;
+  }
+  mooseError("too old history value requested from StepperInfo");
+}
+
+Real
+StepperInfo::dt(int n)
+{
+  return nth_elem(_dt, n);
```

Comment:
I'm not crazy about this name since it exists in the standard namespace and doesn't something different than your implementation. You might want to reconsider the name.